dom.bs

```diff
@@ -4532,33 +4532,86 @@ each <a for=tree>descendant</a> <a>exclusive <code>Text</code> node</a> <var>nod
-<var>node</var>, with an optional <var>document</var> and <i>clone children flag</i>, run these
-steps:
-<!-- This algorithm is used by dom-Node-cloneNode, dom-Document-importNode,
-dom-Range-extractContents, dom-Range-cloneContents -->
+<div algorithm>
+<p>To <dfn export id=concept-node-clone local-lt="clone">clone a node</dfn> given a
+<a for=/>node</a> <var>node</var> and an optional <a for=/>document</a> <var>document</var> (default
+<var>node</var>'s <a for=Node>node document</a>) and boolean <var>subtree</var> (default false):
 
 <ol>
- <li><p>If <var>document</var> is not given, then set <var>document</var> to <var>node</var>'s
- <a for=Node>node document</a>.
+ <li><p><a for=/>Assert</a>: <var>node</var> is not a <a for=/>document</a> or <var>node</var> is
+ <var>document</var>.
+
+ <li><p>Return the result of <a>cloning a node and append</a> given <var>node</var>, null,
+ <var>document</var>, and <var>subtree</var>.
+</ol>
+</div>
+
+<div algorithm>
+<p>To <dfn export>clone a node and append</dfn> given a <a for=/>node</a> <var>node</var>, a
+<a for=/>node</a>-or-null <var>parent</var>, a <a for=/>document</a> <var>document</var>, and a
+boolean <var>subtree</var>:
 
- <li><a for=/>Assert</a>: <var>node</var> is not a <a for=/>document</a> or <var>node</var> is
+<ol>
+ <li><p>Let <var>copy</var> be the result of <a>cloning a single node</a> given <var>node</var> and
  <var>document</var>.
 
+ <li><p>If <var>parent</var> is non-null, then <a for=/>append</a> <var>copy</var> to
+ <var>parent</var>.
+
+ <li><p>Run any <a>cloning steps</a> defined for <var>node</var> in
+ <a>other applicable specifications</a> and pass <var>node</var>, <var>copy</var>, and
+ <var>subtree</var> as parameters.
+
+ <li>
+  <p>If <var>node</var> is an <a for=/>element</a>, <a for=Element>shadow host</a>, and its
+  <a for=Element>shadow root</a>'s <a for=ShadowRoot>clonable</a> is true:
+
+  <ol>
+   <li><p><a for=/>Assert</a>: <var>copy</var> is not a <a for=Element>shadow host</a>.
+
+   <li><p><a>Attach a shadow root</a> with <var>copy</var>, <var>node</var>'s
+   <a for=Element>shadow root</a>'s <a for=ShadowRoot>mode</a>, true, <var>node</var>'s
+   <a for=Element>shadow root</a>'s <a for=ShadowRoot>serializable</a>, <var>node</var>'s
+   <a for=Element>shadow root</a>'s <a for=ShadowRoot>delegates focus</a>, and <var>node</var>'s
+   <a for=Element>shadow root</a>'s <a for=ShadowRoot>slot assignment</a>.
+
+   <li><p>Set <var>copy</var>'s <a for=Element>shadow root</a>'s <a for=ShadowRoot>declarative</a>
+   to <var>node</var>'s <a for=Element>shadow root</a>'s <a for=ShadowRoot>declarative</a>.
+
+   <li><p>For each <var>child</var> of <var>node</var>'s <a for=Element>shadow root</a>'s
+   <a for=tree>children</a>, in <a>tree order</a>: <a>clone a node and append</a> given
+   <var>child</var>, <var>copy</var>'s <a for=Element>shadow root</a>, <var>document</var>, and
+   <var>subtree</var>.
```

Comment:
In Gecko shadow root attaching happens after going through children.
And the same seems to happen in Chrome too https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=689;drc=fe99dee25a6b65dbe6c9e97099334e9c027a24a1;bpv=0;bpt=1

The order of slotchange events is the same in Firefox and Chrome (h6, div, span), Safari has - I was told - div, span, h6.

Reply:
Yeah it is, but is that the result of cloning order or something else? As per the discussion in #695 there's already a difference between engines.

I can restore the original order here though. That seems better for now either way.